sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,1121 @@
+            ((Configurable) inputFormatObj).setConf(conf.getHadoopConfiguration());
+          }
+        } catch (InstantiationException e) {
+          throw new IOException("Unable to create InputFormat object: ", e);
+        } catch (IllegalAccessException e) {
+          throw new IOException("Unable to create InputFormat object: ", e);
+        } catch (ClassNotFoundException e) {
+          throw new IOException("Unable to create InputFormat object: ", e);
+        }
+      }
+    }
+
+    /**
+     * Throws exception if you set different InputFormat key or value class than InputFormat's
+     * actual key or value class. If you set incorrect classes then, it may result in an error like
+     * "unexpected extra bytes after decoding" while the decoding process happens. Hence this
+     * validation is required.
+     */
+    private void validateUserInputForKeyAndValue() throws IOException, InterruptedException {
+      ParameterizedType genericClassType = determineGenericType();
+      boolean isCorrectKeyClassSet = validateClasses(
+          genericClassType.getActualTypeArguments()[0].getTypeName(),
+          keyCoder,
+          "key.class");
+      boolean isCorrectValueClassSet = validateClasses(
+          genericClassType.getActualTypeArguments()[1].getTypeName(),
+          valueCoder,
+          "value.class");
+      if (!isCorrectKeyClassSet) {
+        Class<?> actualClass = conf.getHadoopConfiguration().getClass("key.class", Object.class);
+        throw new IllegalArgumentException(String.format(
+            HadoopInputFormatIOConstants.WRONG_INPUTFORMAT_KEY_CLASS_ERROR_MSG,
+            getExpectedKeyClass().getName(),
+            actualClass.getName()));
+      }
+      if (!isCorrectValueClassSet) {
+        Class<?> actualClass = conf.getHadoopConfiguration().getClass("value.class", Object.class);
+        throw new IllegalArgumentException(String.format(
+            HadoopInputFormatIOConstants.WRONG_INPUTFORMAT_VALUE_CLASS_ERROR_MSG,
+            getExpectedValueClass().getName(),
+            actualClass.getName()));
+      }
+    }
+
+    /**
+     * Returns true if key/value class set by the user is compatible with the key/value class of a
+     * pair returned by RecordReader. User provided key and value classes are validated against
+     * parameterized type of InputFormat. If parameterized type has any type parameter such as T,
+     * K, V, etc then validation is done by encoding and decoding first pair returned by
+     * RecordReader.
+     */
+    private <T> boolean validateClasses(
+        String inputFormatGenericType,
+        Coder<T> coder,
+        String property) throws IOException, InterruptedException {
+      Class<?> inputClass = null;
+      try {
+        inputClass = Class.forName(inputFormatGenericType);
+      } catch (Exception e) {
+        /*
+         * Given inputFormatGenericType is a type parameter i.e. T, K, V, etc. In such cases class
```

Comment:
I don't understand this comment - can you explain this a bit more in the comment? Why does the fact that it's a type parameter mean that we need to validate it using the coder? 

Also, I don't understand what you mean when you say that it's a "type parameter" - all the T/K/V are forgotten at run time, so what we have here is not actually those value.